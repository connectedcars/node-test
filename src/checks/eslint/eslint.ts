import { CheckAnnotation, CheckAnnotationLevel, CheckConversionError, CheckRunCompleted } from '../checks-common'
import { EslintData } from './eslint-types'

export interface EslintInput {
  sha: string
  data: EslintData[]
}

const generateSummary = (errors: number, warnings: number): string => {
  const problems = errors + warnings
  if (!problems) {
    return 'No problems found'
  }

  const summary = `Found ${problems} ${problems === 1 ? 'problem' : 'problems'}`

  const details = []
  if (errors > 0) {
    details.push(`${errors} ${errors === 1 ? 'error' : 'errors'}`)
  }

  if (warnings > 0) {
    details.push(`${warnings} ${warnings === 1 ? 'warning' : 'warnings'}`)
  }

  return `${summary} (${details.join(', ')})`
}

export const eslintCheck = ({ data, sha }: EslintInput): CheckRunCompleted => {
  try {
    let errors = 0
    let warnings = 0
    const annotations: CheckAnnotation[] = []
    // Run through each file
    outer: for (const file of data) {
      // Run through each message for file
      for (const message of file.messages) {
        if (message.message === 'File ignored because of a matching ignore pattern. Use "--no-ignore" to override.') {
          continue outer
        }
        const match = file.filePath.match(/^.*\/((?:src|bin)\/.+)$/)
        const relPath = match && match.length === 2 ? match[1] : '.eslintrc'
        // Determine severity of message
        let annotation_level: CheckAnnotationLevel = 'notice'
        switch (message.severity) {
          case 1:
            annotation_level = 'notice'
            break
          case 2:
            annotation_level = 'failure'
            break
        }
        // Generate an annotation
        const annotation: CheckAnnotation = {
          path: relPath,
          annotation_level,
          start_line: message.line || 1,
          end_line: message.endLine || 1,
          message: `${message.line}:${message.column}`.padEnd(10) + message.message,
          raw_details: JSON.stringify(message, null, '    ')
        }

        annotations.push(annotation)
      }
      // Increment problem counts
      errors += file.errorCount
      warnings += file.warningCount
    }

    const summary = generateSummary(errors, warnings)

    const result: CheckRunCompleted = {
      name: 'eslint',
      head_sha: sha,
      conclusion: 'success',
      status: 'completed',
      completed_at: new Date().toISOString(),
      output: {
        title: summary,
        summary,
        // Limit to 50 annotations as this is the max per post for github
        annotations: annotations.slice(0, 50)
      }
    }

    if (errors > 0) {
      result.conclusion = 'failure'
      return result
    }

    if (warnings > 0) {
      result.conclusion = 'neutral'
      return result
    }

    return result
  } catch (e) {
    throw new CheckConversionError('eslint', { data, sha }, e)
  }
}
