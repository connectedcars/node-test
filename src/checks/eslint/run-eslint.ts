import { runJsonCommand } from '../checks-common'

export async function runEslint(args: string[] = []): Promise<EslintData[]> {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const [exitInfo, json] = await runJsonCommand<EslintData[]>('eslint', [
    '--format',
    'json',
    ...(args.length > 0 ? args : ['./src/**/*.{ts,tsx}'])
  ])

  return json
}
