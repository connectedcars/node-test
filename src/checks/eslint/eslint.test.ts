import { eslintCheck } from './eslint'
import {
  eslintErrorAnnotationsOutput,
  eslintErrorOutput,
  eslintSkippedOutput,
  eslintSuccessfulOutput
} from './resources/eslint-help-text'

describe('checks/eslint', () => {
  it('converts successful eslint', () => {
    const data = JSON.parse(eslintSuccessfulOutput)
    const result = eslintCheck({
      data,
      org: 'connectedcars',
      repo: 'cloudbuilder-wrapper',
      sha: 'c61a4ae014360e064eb2a9f76c8a6a55d05e5b88'
    })
    expect(result).toMatchSnapshot({
      completed_at: expect.stringMatching(/^\d{4}/)
    })
  })

  it('converts successful failed eslint', () => {
    const data = JSON.parse(eslintErrorOutput)
    const result = eslintCheck({
      data,
      org: 'connectedcars',
      repo: 'cloudbuilder-wrapper',
      sha: 'c61a4ae014360e064eb2a9f76c8a6a55d05e5b88'
    })
    expect(result).toMatchSnapshot({
      completed_at: expect.stringMatching(/^\d{4}/)
    })
  })

  it('converts successful failed eslint with annotations', () => {
    const data = JSON.parse(eslintErrorAnnotationsOutput)
    const result = eslintCheck({
      data,
      org: 'connectedcars',
      repo: 'cloudbuilder-wrapper',
      sha: 'c61a4ae014360e064eb2a9f76c8a6a55d05e5b88'
    })
    expect(result).toMatchSnapshot({
      completed_at: expect.stringMatching(/^\d{4}/)
    })
  })

  it('converts successful skipped eslint', () => {
    const data = JSON.parse(eslintSkippedOutput)
    const result = eslintCheck({
      data,
      org: 'connectedcars',
      repo: 'cloudbuilder-wrapper',
      sha: 'c61a4ae014360e064eb2a9f76c8a6a55d05e5b88'
    })
    expect(result).toMatchSnapshot({
      completed_at: expect.stringMatching(/^\d{4}/)
    })
  })
})
