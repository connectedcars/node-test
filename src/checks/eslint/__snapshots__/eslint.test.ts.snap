// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checks/eslint converts successful eslint 1`] = `
Object {
  "completed_at": StringMatching /\\^\\\\d\\{4\\}/,
  "conclusion": "success",
  "head_sha": "c61a4ae014360e064eb2a9f76c8a6a55d05e5b88",
  "name": "eslint",
  "output": Object {
    "annotations": Array [],
    "summary": "No problems found",
    "title": "No problems found",
  },
  "status": "completed",
}
`;

exports[`checks/eslint converts successful failed eslint 1`] = `
Object {
  "completed_at": StringMatching /\\^\\\\d\\{4\\}/,
  "conclusion": "failure",
  "head_sha": "c61a4ae014360e064eb2a9f76c8a6a55d05e5b88",
  "name": "eslint",
  "output": Object {
    "annotations": Array [],
    "summary": "Found 1 problem (1 error)",
    "title": "Found 1 problem (1 error)",
  },
  "status": "completed",
}
`;

exports[`checks/eslint converts successful failed eslint with annotations 1`] = `
Object {
  "completed_at": StringMatching /\\^\\\\d\\{4\\}/,
  "conclusion": "failure",
  "head_sha": "c61a4ae014360e064eb2a9f76c8a6a55d05e5b88",
  "name": "eslint",
  "output": Object {
    "annotations": Array [
      Object {
        "annotation_level": "failure",
        "end_line": 26,
        "message": "26:1      Delete \`↹\`",
        "path": "src/checks/jest/jest.ts",
        "raw_details": "{
    \\"ruleId\\": \\"prettier/prettier\\",
    \\"severity\\": 2,
    \\"message\\": \\"Delete \`↹\`\\",
    \\"line\\": 26,
    \\"column\\": 1,
    \\"nodeType\\": null,
    \\"endLine\\": 26,
    \\"endColumn\\": 2,
    \\"fix\\": {
        \\"range\\": [
            679,
            680
        ],
        \\"text\\": \\"\\"
    }
}",
        "start_line": 26,
      },
      Object {
        "annotation_level": "failure",
        "end_line": 28,
        "message": "27:4      Delete \`⏎\`",
        "path": "src/checks/jest/jest.ts",
        "raw_details": "{
    \\"ruleId\\": \\"prettier/prettier\\",
    \\"severity\\": 2,
    \\"message\\": \\"Delete \`⏎\`\\",
    \\"line\\": 27,
    \\"column\\": 4,
    \\"nodeType\\": null,
    \\"endLine\\": 28,
    \\"endColumn\\": 1,
    \\"fix\\": {
        \\"range\\": [
            701,
            702
        ],
        \\"text\\": \\"\\"
    }
}",
        "start_line": 27,
      },
    ],
    "summary": "Found 2 problems (2 errors)",
    "title": "Found 2 problems (2 errors)",
  },
  "status": "completed",
}
`;

exports[`checks/eslint converts successful skipped eslint 1`] = `
Object {
  "completed_at": StringMatching /\\^\\\\d\\{4\\}/,
  "conclusion": "success",
  "head_sha": "c61a4ae014360e064eb2a9f76c8a6a55d05e5b88",
  "name": "eslint",
  "output": Object {
    "annotations": Array [],
    "summary": "No problems found",
    "title": "No problems found",
  },
  "status": "completed",
}
`;
